File: 2015/day13.py
from itertools import permutations

data = open("day13_input.txt").readlines()
h = {}
ppl = []

for line in data:
  parts = line.split(' ')

  a = parts[0]
  if parts[2] == 'lose':
    units = -int(parts[3])
  else:
    units = int(parts[3])
  b = parts[-1]

  h[a + b] = units

  if a not in ppl: ppl.append(a)

print(h)
print(ppl)

perms = list(permutations(ppl))

part1 = 0

for perm in perms:
  total = 0
  for x in range(len(ppl)):
    if x == (len(ppl)-1):
      y = 0
    else:
      y = x + 1
    a = perm[x] + perm[y]
    # print("weight for", a, ":", weights[a])
    b = perm[y] + perm[x]
    # print("weight for", b, ":", weights[b])
    total += h[a] + h[b]

  print("Total for",perm," = ", total)
  if total > part1: part1 = total

print('Part 1:', part1)

File: 2015/Day13.rb
h = {}

me = 'Craig'

File.readlines('Day13_input.txt').each do |line|
  parts = line.chomp[0..-2].split(' ')
  a = parts[0]
  units = if parts[2] == 'gain'
            parts[3].to_i
          else
            -parts[3].to_i
          end
  b = parts[-1]
  # puts "#{a}#{b}:#{units}"
  # h[a + b] = units
  (h[a] ||= {})[b] = units

  # part 2
  if me != ''
    (h[a] ||= {})[me] = 0
    (h[me] ||= {})[a] = 0
  end
end

t = h.keys.permutation.map { |p|
  (p << p[0]).each_cons(2).map { |a, b|
    h[a][b] + h[b][a]
  }.reduce(:+)
  # puts "Perm: #{p} = #{t}"
}

if me == ''
  puts "Part 1: #{t.max}"
else
  puts "Part 2: #{t.max}"
end
